File: exercism/javascript/linked-list/linked-list.js
const LinkedList = () => {

  let head = null
  let tail = null

  const Node = (data, next, previous) => {
    return { data, next, previous }
  }

  const count = () => {
    let counter = head
    let listSize = 0
    while (counter != null && counter.data != null) {
      listSize++
      counter = counter.next
    }
    return listSize
  }

  const push = (obj) => {
    let newNode = new Node(obj, null, null)
    if (!head) {
      head = newNode
      tail = newNode
    } else {
      newNode.previous = tail
      tail.next = newNode
      tail = newNode
    }
  }

  const pop = () => {
    let output = tail.data;
    if (!head) {
      throw (new Error ("Can not pop an empty list"))
    } else if (tail != head) {
      tail = tail.previous
      tail.next = null
    } else {
      output = head.data
      head = null
      tail = null
    }
    return output
  }

  const shift = () => {
    let output = head.data;
    if (!head) {
      throw (new Error ("Can not shift an empty list"))
    } else if (head != tail) {
      head = head.next
      head.previous = null
    } else {
      output = tail.data
      head = null
      tail = null
    }
    return output
  }

  const unshift = (obj) => {
    let newNode = new Node(obj, null, null)
    if (!head) {
      head = newNode
      tail = newNode
    } else {
      newNode.next = head
      head.previous = newNode
      head = newNode
    }
  }

  const deleteValue = (obj) => {

  }

  return { push, pop, shift, unshift, count, deleteValue }
}

export default LinkedList

File: exercism/ruby/complete/linked-list/linked_list.rb
class Node
  attr_accessor :data, :previous, :next

  def initialize(data)
    @data = data
    @previous = nil
    @next = nil
  end
end

class Deque
  def initialize
    @head = nil
    @tail = nil
  end

  def push(data)
    new_node = Node.new(data)
    if @head.nil? && @tail.nil?
      @head = new_node
      @tail = new_node
    else
      @tail.next = new_node
      new_node.previous = @tail
      @tail = new_node
    end
    new_node
  end

  def pop
    output = @tail.data
    unless @tail.previous.nil?
      @tail = @tail.previous
      @tail.next = nil
    else
      @head = nil
      @tail = nil
    end
    output
  end

  def shift
    output = @head.data
    unless @head.next.nil?
      @head = @head.next
      @head.previous = nil
    else
      @head = nil
      @tail = nil
    end
    output
  end

  def unshift(data)
    new_node = Node.new(data)
    if @head.nil? && @tail.nil?
      @head = new_node
      @tail = new_node
    else
      @head.previous = new_node
      new_node.next = @head
      @head = new_node
    end
    new_node
  end
end
